From: Stanislav Malyshev <stas@php.net>
Date: Mon, 27 May 2019 17:16:29 -0700
Subject: Fix bug #77988 - heap-buffer-overflow on php_jpg_get16

(cherry picked from commit 73ff4193be24192c894dc0502d06e2b2db35eefb)
---
 NEWS                         |  14 ++++++++++++++
 ext/exif/exif.c              |   2 ++
 ext/exif/tests/bug77988.jpg  | Bin 0 -> 1202 bytes
 NEWS                         |   14 ++++++++++++++
 ext/exif/exif.c              |    2 ++
 ext/exif/tests/bug77988.jpg  |binary
 NEWS                         |   14 ++++++++++++++
 ext/exif/exif.c              |    2 ++
 ext/exif/tests/bug77988.phpt |   11 +++++++++++
 3 files changed, 27 insertions(+)
 create mode 100644 ext/exif/tests/bug77988.jpg
 create mode 100644 ext/exif/tests/bug77988.phpt

--- a/NEWS
+++ b/NEWS
@@ -1,6 +1,20 @@
 PHP                                                                        NEWS
 |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
 
+Backported from 7.1.30
+
+- EXIF:
+  . Fixed bug #77988 (heap-buffer-overflow on php_jpg_get16).
+    (CVE-2019-11040) (Stas)
+
+- GD:
+  . Fixed bug #77973 (Uninitialized read in gdImageCreateFromXbm).
+    (CVE-2019-11038) (cmb)
+
+- Iconv:
+  . Fixed bug #78069 (Out-of-bounds read in iconv.c:_php_iconv_mime_decode()
+    due to integer overflow). (CVE-2019-11039). (maris dot adam)
+
 Backported from 7.1.29
 
 - EXIF
--- a/ext/exif/exif.c
+++ b/ext/exif/exif.c
@@ -3536,6 +3536,8 @@ static int exif_scan_thumbnail(image_inf
 		if (c == 0xFF)
 			return FALSE;
 		marker = c;
+		if (pos>=ImageInfo->Thumbnail.size)
+			return FALSE;
 		length = php_jpg_get16(data+pos);
 		if (length > ImageInfo->Thumbnail.size || pos >= ImageInfo->Thumbnail.size - length) {
 			return FALSE;
--- /dev/null
+++ b/ext/exif/tests/bug77988.phpt
@@ -0,0 +1,11 @@
+--TEST--
+Bug #77988 (heap-buffer-overflow on php_jpg_get16)
+--SKIPIF--
+<?php if (!extension_loaded('exif')) print 'skip exif extension not available';?>
+--FILE--
+<?php
+exif_read_data(__DIR__."/bug77988.jpg", 'COMMENT', FALSE, TRUE);
+?>
+DONE
+--EXPECTF--
+DONE
\ No newline at end of file
